File: Maximum Depth of B-Tree/Python/solution1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        return self.max(root, 1, 1)
        
    def max(self, root, maxv, level):
        if root is None:
            return level-1
        # Recurse the left side. If the depth is greater than the tracked max set it to the max
        depth = self.max(root.left, maxv, level+1)
        if depth > maxv:
            maxv = depth
        # Recurse the right side. If the depth is greater than the tracked max set it to the max
        depth = self.max(root.right, maxv, level+1)
        if depth > maxv:
            maxv = depth
            
        return maxv

File: Maximum Depth of B-Tree/Java/solution1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int maxDepth(TreeNode root) {
        // If root is null return 0 as there is not node
        if(root == null){
           return 0;
       }
       // get the max from the left side then the max from the right
       // Add 1 for each recursive call
       return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }
}